plutus-core/untyped-plutus-core/src/UntypedPlutusCore/NewParser.hs

```diff
@@ -0,0 +1,146 @@
+appTerm tm = UPLC.Apply <$> getSourcePos <*> tm <*> tm
+
+conTerm
+    :: (PLC.Parsable (PLC.Some uni), PLC.Closed uni, uni `PLC.Everywhere` PLC.Parsable)
+    => Parser (UPLC.Term PLC.Name uni fun SourcePos)
+conTerm = UPLC.Constant <$> reservedWord "con" <*> constant
+
+builtinTerm :: (Bounded fun, Enum fun, Pretty fun) => Parametric name uni fun
+builtinTerm _term = UPLC.builtin <$> reservedWord "builtin" <*> builtinFunction
+
+tyInstTerm :: ParsecT ParseError
+                   T.Text
+                   (StateT ParserState PLC.Quote)
+                   (UPLC.Term PLC.Name uni fun SourcePos)
+                   -> Parser (UPLC.Term PLC.Name uni fun SourcePos)
+tyInstTerm tm = UPLC.Force <$> getSourcePos <*> tm
+
+-- In uplc, Iwrap and Unwrap are removed.
+
+errorTerm
+    :: Parser (UPLC.Term PLC.Name uni fun SourcePos)
+errorTerm = UPLC.Error <$> reservedWord "error"
+
+term
+    :: ( PLC.Parsable (PLC.Some uni), PLC.Closed uni, uni `PLC.Everywhere` PLC.Parsable
+       , Bounded fun, Enum fun, Pretty fun
+       )
+        => Parser (UPLC.Term PLC.Name uni fun SourcePos)
+term = (var >>= (\n -> getSourcePos >>= \p -> return $ UPLC.Var p n))
+    <|> (inParens $ absTerm self <|> lamTerm self <|> conTerm <|> builtinTerm self <|> errorTerm)
+    <|> inBraces (tyInstTerm self)
+    <|> inBrackets (appTerm self)
+    where self = term
+
+program
+    :: ( PLC.Parsable (PLC.Some uni), PLC.Closed uni, uni `PLC.Everywhere` PLC.Parsable
+       , Bounded fun, Enum fun, Pretty fun
+       )
+    => Parser (UPLC.Program PLC.Name uni fun SourcePos)
+program = whitespace >> do
+    prog <- inParens $ UPLC.Program <$> reservedWord "program" <*> version <*> term
+    notFollowedBy anySingle
+    return prog
+
+-- | Generic parser function
+parseGen :: Parser a -> ByteString -> Either (ParseErrorBundle T.Text ParseError) a
+parseGen stuff bs = parse stuff "test" $ (T.pack . unpackChars) bs
+
+-- | Parse a PLC term. The resulting program will have fresh names. The underlying monad must be capable
+-- of handling any parse errors.
+parseTerm :: (PLC.Parsable (PLC.Some uni), PLC.Closed uni, uni `PLC.Everywhere` PLC.Parsable
+       , Bounded fun, Enum fun, Pretty fun
+       ) => ByteString -> Either (ParseErrorBundle T.Text ParseError) (UPLC.Term PLC.Name uni fun SourcePos)
+parseTerm = parseGen term
+
+-- | Parse a PLC program. The resulting program will have fresh names. The underlying monad must be capable
+-- of handling any parse errors.
+parseProgram :: (PLC.Parsable (PLC.Some uni), PLC.Closed uni, uni `PLC.Everywhere` PLC.Parsable
+       , Bounded fun, Enum fun, Pretty fun
+       ) => ByteString -> Either (ParseErrorBundle T.Text ParseError) (UPLC.Program PLC.Name uni fun SourcePos)
+parseProgram = parseGen program
+
+-- | Take one UPLC program and apply it to another.
+applyProgram :: UPLC.Program name uni fun () -> UPLC.Program name uni fun () -> UPLC.Program name uni fun ()
```

Comment:
this exists already, and isn't really related to the parser?

Reply:
Oh yes, it was in `UntypedPlutusCore` which I will replace, but I should put it in `NewUntypedPlutusCore`.